### PR TITLE
docs: custom ui

### PR DESCRIPTION
**Pull request by Benehiko**
## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md) and signed the CLA.
- [ ] I have referenced an issue containing the design document if my change introduces a new feature.
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security vulnerability. 
      If this pull request addresses a security vulnerability, 
      I confirm that I got approval (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added the necessary documentation within the code base (if appropriate).

### CODE REVIEW

File: docs/concepts/custom-ui-headless-api.mdx [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+For more information on the concepts, take a look at the [Self-service flow](../kratos/self-service) documentation.
+
+## Creating a flow
+
+All browser flows can be initialized with the optional `?return_to=<url>` query parameter. This parameter is used to redirect the
+user to the specified URL after the flow is successfully completed.
+
+Some flows, such as login have additional query parameters which can be used in both browser and native applications.
+
+## Reusing an existing flow
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:36]: I am not very happy with this section. What I think is confusing is:

- Wording: we are not "resuing" the flow, but rather fetching it. The reader might think they can create a flow and use it for multiple subsequent logins.
- The section should be further down when some of the more fundamental concepts are explained.
- The text is too generic. It is a complete mystery at this point why "Ory has initialized the flow and redirects to the UI". This might be more clear if that is explained in the "Create" section.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+
+This documents' focus is to provide concrete examples on custom UI integration. Flows are the fundamental of Ory's self-service
+APIs and are used to accomplish a variety of goals, such as `login`, `registration`, `settings`, `verification` and `recovery`.
+
+The self-service `flows` are standardized so that it follows the same steps regardless of which `flow` is integrated.
+
+A general overview of the flow can be summed up in two steps:
+
+1. Get or create the `flow` data, e.g. login, registration etc.
+2. Submit the `flow` with user data (e.g. username and password)
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:25]: I would make this more verbose:

```suggestion
1. Create the flow with the appropriate purpose and application.
2. Use the flow data to render the UI.
3. Submit the flow with user data.
4. Potentially handle flow errors like invalid values by going back to 2.
5. Handle successful submission.
```

The use-cases can become very complex, but first explaining the "standard" use-case should help people to wrap their head around the concepts.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+
+For more information on the concepts, take a look at the [Self-service flow](../kratos/self-service) documentation.
+
+## Creating a flow
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:29]: In the end this section does not even explain how to create a flow, but only what special params are there.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+Here are some cases where the flow would already exist:
+
+1. Ory has initialized the flow and redirects to the UI (e.g. `/login?flow=<uuidv4>`)
+2. The request was completed by the user, however, an error occurred, and we need to retrieve the flow data to display the error
+3. The UI has 'stored' the flow ID in the URL so that page refreshes can retrieve the already existing flow data
+
+:::note
+
+The flow ID has an expiration time, so it is important to check the response status code when retrieving the flow data using an
+existing flow ID. It is up to your application to handle this error case and create a new flow.
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:51]: Error handling should be it's own section. There is also two kinds of errors: validation and all others.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+
+:::
+
+## Submitting a flow
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:55]: This section does also not explain how to submit a flow, and what happens after that. This should be added IMO.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+Never use API flows to implement Browser applications! Using API flows in Single-Page-Apps as well as server-side apps opens up
+several potential attack vectors, including Login and other CSRF attacks.
+
+:::
+
+## Login flow
+
+Login flows are used to authenticate users. When the login is successful, the request will respond with a `200 OK` HTTP status
+code and a JSON body containing the user's identity information. When the request fails, it responds with the appropriate HTTP
+status and error messages.
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:72]: This is not really true. After successful login the server issues a session depending on the application (cookie or token).

File: docs/concepts/custom-ui-headless-api.mdx [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+
+#### Two-Factor authentication
+
+Two-Factor authentication (2FA) is a second step to confirming the user's credentials with another credential, such as `TOTP` or
+`TouchID`. When creating a login flow, you can specify the `?aal=<level>` query parameter to specify the level of authentication.
+
+In Ory the levels are:
+
+- aal1: Password or OIDC
+- aal2: aal1 + TOTP or WebAuthn or Lookup Secrets
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:124]: Maybe instead link to `../mfa/overview#authenticator-assurance-level-aal`

File: docs/concepts/custom-ui-headless-api.mdx [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+<CodeBlock language="ts">{createReactLoginApi}</CodeBlock>
+```
+
+</TabItem>
+
+
+</Tabs>
+````
+
+#### Two-Factor authentication
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:116]: We should add to this section that an login flow is also used to upgrade an `aal1` to an `aal2` session.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+
+In Ory the levels are:
+
+- aal1: Password or OIDC
+- aal2: aal1 + TOTP or WebAuthn or Lookup Secrets
+
+In the case of initializing a second factor, the user must already have a valid session cookie. The
+[`/sessions/whoami`](../reference/api#tag/frontend/operation/toSession) endpoint will return with an error with an id
+`session_aal2_required` if the user is required to complete a second factor. More information on UI error messages can be found on
+the [Customize user interface](../kratos/concepts/ui-user-interface) page.
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:129]: This seems a bit irrelevant for understanding login flows.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+
+</Tabs>
+````
+
+### Submit Login flow
+
+We can submit the login flow with user data (e.g. username and password) using the flow ID. Depending on the type of application
+(e.g. browser or native) the CSRF tokens are required in the request body and cookie.
+
+In some cases, such as submitting WebAuthn passwordless, the response code could be `422` which indicates that the browser needs
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:176]: I think 422 is a typo here.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+to be redirected with a newly created flow. In Single Page Applications (SPAs) instead of redirecting, the response can be decoded
+and the new flow ID extracted.
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:178]: This concept of JSON payload vs redirect is true for basically all API calls and should be explained way earlier, not in this subsubsection.

File: docs/concepts/custom-ui-headless-api.mdx [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+```mdx-code-block
+import Tabs from "@theme/Tabs"
+import TabItem from "@theme/TabItem"
+import CodeBlock from "@theme/CodeBlock"
+```
+
+Ory provides headless APIs for ease of integration. This ensures that Ory provides compatibility with many software stacks across
+different programming languages.
+
+This documents' focus is to provide concrete examples on custom UI integration. Flows are the fundamental of Ory's self-service
```
Comment (zepatrik) [inline docs/concepts/custom-ui-headless-api.mdx:17]: In the end, it is not really clear how to render a form from a flow, or when to create a flow in the UI in the first place. The page does really answer the question posed here.

**Review by zepatrik — DISMISSED** [review]
I don't want to sound to harsh on this review, it is a lot easier to just point at something instead of writing a good docs page, but I think the page needs a bit more work.

**Comment by vinckr** [comment]
huge PR 🚀 

should I do a review @Benehiko ? not everything but I could try out the react examples at least?

**Comment by Benehiko** [comment]
> huge PR 🚀 
> 
> should I do a review @Benehiko ? not everything but I could try out the react examples at least?

That would be much appreciated:)
I think it is difficult to get code right inside such a document and it will be hard to maintain. So we should have somewhere here a disclaimer that although we try make it as copy pastable as possible, it isn't always the case.

**Review by vinckr — COMMENTED** [review]
some suggestions for the text, feel free to disregard @tomekpapiernik but could make it a bit more readable. 
I will now just concentrate on the code part.

**Review by vinckr — COMMENTED** [review]
I feel the basic react example would be a bit more basic without ory/integrations.

//edit: Or maybe not, can you just use the ory/integrations UI Node Helpers in general (in react) as helper functions? Seems really useful for building a custom UI, would even merit their own doc in that case IMO.

**Comment by Benehiko** [comment]
> I feel the basic react example would be a bit more basic without ory/integrations.
> 
> //edit: Or maybe not, can you just use the ory/integrations UI Node Helpers in general (in react) as helper functions? Seems really useful for building a custom UI, would even merit their own doc in that case IMO.

I mean they could be more basic by referencing the traits and things directly, instead of dynamically. But I thought this might be the best to show dynamic mapping without too many libraries. using the Ui Node helpers is that sweet spot between directly referencing the traits and ory/elements. So yes you can and yest it would be good to have docs on it.

In general I would think our users would be in 3 categories:
1. Use Account Experience with good theming and customization
2. Built it yourself with ory elements + good theming and ability to use own components (e.g. button and input box)
3. Build everything yourself - in this case you could use ory/integrations but sometimes that even wouldn't make sense (i only need passwordless and no dynamic UI).

**Review by tomekpapiernik — APPROVED** [review]
ooofff
